File: 173.二叉搜索树迭代器.go
/*
 * @lc app=leetcode.cn id=173 lang=golang
 *
 * [173] 二叉搜索树迭代器
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

 /*
//  以前通过版
type BSTIterator struct {
	stack []*TreeNode
}


func Constructor(root *TreeNode) BSTIterator {
	stack := []*TreeNode{}
	if root != nil {
		stack = append(stack,root)
		for root.Left != nil {
			stack = append(stack,root.Left)
			root = root.Left
		}
	}
	return BSTIterator{stack}
}


func (this *BSTIterator) Next() int {
	l := len(this.stack)
	node := this.stack[l-1]
	ans := node.Val

	if node.Right != nil {
		node = node.Right
		this.stack[l-1] = node
		for node.Left != nil {
			node = node.Left
			this.stack = append(this.stack,node)
		}
	}else{
		this.stack = this.stack[:l-1]
	}
	return ans
}


func (this *BSTIterator) HasNext() bool {
	return len(this.stack) > 0
}
*/


// 20210328每日一题版
type BSTIterator struct {
	nodes []*TreeNode
}

func Constructor(root *TreeNode) BSTIterator {
	nodes := []*TreeNode{}
	if root != nil {
		nodes = append(nodes,root)
		for root.Left != nil {
			nodes = append(nodes,root.Left)
			root = root.Left
		}
	}
	return BSTIterator{nodes}
}

func (this *BSTIterator) Next() int {
	l := len(this.nodes)
	node := this.nodes[l-1]
	val := node.Val
	
	this.nodes = this.nodes[:l-1]

	if node.Right != nil {
		this.nodes = append(this.nodes,node.Right)
		node = node.Right
		for node.Left != nil {
			this.nodes = append(this.nodes,node.Left)
			node = node.Left
		}
	}
	return val
}

func (this *BSTIterator) HasNext() bool {
	return len(this.nodes) > 0
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Next();
 * param_2 := obj.HasNext();
 */
// @lc code=end

File: 173.二叉搜索树迭代器.ts
/*
 * @lc app=leetcode.cn id=173 lang=typescript
 *
 * [173] 二叉搜索树迭代器
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
// 二叉树的中序遍历
class BSTIterator {
    nodes: TreeNode[] = new Array<TreeNode>();

    constructor(root: TreeNode | null) {
        let cur:TreeNode = root;
        for (;cur !== null;) {
            this.nodes.push(cur);
            cur = cur.left;
        }
    }

    next(): number {
        let node: TreeNode = this.nodes.pop();
        let res: number = node.val;
        if (node.right !== null) {
            node = node.right;
            for (;node !== null;) {
                this.nodes.push(node);
                node = node.left;
            }
        }
        return res;
    }

    hasNext(): boolean {
        return this.nodes.length > 0;
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = new BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */
// @lc code=end

